src/datachain/catalog/catalog.py

```diff
@@ -509,6 +509,60 @@ def clone_catalog_with_cache(catalog: "Catalog", cache: "Cache") -> "Catalog":
     return clone
 
 
+def get_all_ids(d: dict[str, Any]) -> list[int]:
+    """
+    Extracts all unique dataset IDs from a nested dictionary structure.
+    """
+    ids = set()
+    for key, value in d.items():
+        ids.add(int(key))
+        if isinstance(value, dict):
+            ids.update(get_all_ids(value))
+    return sorted(ids)
+
+
+def build_nested_dependencies(
+    dataset_deps: dict[str, DatasetDependency],
+    dependency_structure: dict[str, dict],
+    indirect: bool = True,
+) -> list[DatasetDependency | None]:
+    """
+    Recursively constructs a tree of dataset dependencies based on their relationships,
+    with each dependency containing its own nested dependencies.
+    """
+
+    def build_deps(dep_id, deps_dict):
+        if dep_id not in dataset_deps:
+            return []
+
+        deps = []
+        if dep_id in deps_dict:
+            for child_id in deps_dict[dep_id]:
+                if child_id in dataset_deps:
+                    # Recursively build child dependencies if indirect=True
+                    child_deps = (
+                        build_deps(child_id, deps_dict[dep_id]) if indirect else []
+                    )
+                    child_dep = dataset_deps[child_id]
+                    child_dep.dependencies = child_deps
+                    deps.append(child_dep)
+                else:
+                    deps.append(None)
+        return deps
+
+    result: list[DatasetDependency | None] = []
+    for root_id in dependency_structure:
+        if root_id in dataset_deps:
+            deps = build_deps(root_id, dependency_structure)
+            root_dep = dataset_deps[root_id]
+            root_dep.dependencies = deps
```

Comment:
C: This is the main purpose of this function. This fills the dependencies in nested manner for the dependency structure. See test_dependency.py for example.